homeassistant/components/directv/media_player.py

```diff
@@ -73,155 +54,64 @@
-        self._is_client = device != "0"
-        self._assumed_state = None
-        self._available = False
-        self._enabled_default = enabled_default
-        self._first_error_timestamp = None
-        self._model = None
-        self._receiver_id = None
-        self._software_version = None
+        self._program = None
+        self._state = None
 
-        if self._is_client:
-            self._model = MODEL_CLIENT
-            self._unique_id = device
-
-        if version_info:
-            self._receiver_id = "".join(version_info["receiverId"].split())
-
-            if not self._is_client:
-                self._unique_id = self._receiver_id
-                self._model = MODEL_HOST
-                self._software_version = version_info["stbSoftwareVersion"]
-
-    def update(self):
+    async def async_update(self):
         """Retrieve latest state."""
-        _LOGGER.debug("%s: Updating status", self.entity_id)
-        try:
-            self._available = True
-            self._is_standby = self.dtv.get_standby()
-            if self._is_standby:
-                self._current = None
-                self._is_recorded = None
-                self._paused = None
-                self._assumed_state = False
-                self._last_position = None
-                self._last_update = None
-            else:
-                self._current = self.dtv.get_tuned()
-                if self._current["status"]["code"] == 200:
-                    self._first_error_timestamp = None
-                    self._is_recorded = self._current.get("uniqueId") is not None
-                    self._paused = self._last_position == self._current["offset"]
-                    self._assumed_state = self._is_recorded
-                    self._last_position = self._current["offset"]
-                    self._last_update = (
-                        dt_util.utcnow()
-                        if not self._paused or self._last_update is None
-                        else self._last_update
-                    )
-                else:
-                    # If an error is received then only set to unavailable if
-                    # this started at least 1 minute ago.
-                    log_message = f"{self.entity_id}: Invalid status {self._current['status']['code']} received"
-                    if self._check_state_available():
-                        _LOGGER.debug(log_message)
-                    else:
-                        _LOGGER.error(log_message)
-
-        except RequestException as exception:
-            _LOGGER.error(
-                "%s: Request error trying to update current status: %s",
-                self.entity_id,
-                exception,
-            )
-            self._check_state_available()
+        self._state = await self.dtv.state(self._address)
+        self._available = self._state.available
+        self._is_standby = self._state.standby
+        self._program = self._state.program
```

Comment:
Why extract these instead of reading from `self.state.available` in the available property? 

Also, what if the request fails or times out ? Can we mark the entity as unavailable?

Reply:
the underlying library is now handling failures. The state object will always be returned error or not. The available property of the state is used to note connectivity to slave receiver unit. Ie is False if connection issue